File: Behavioral/template-method/examples/name-generator/Python/NameGenerator/generators/BaseNameGenerator.py
from enum import Enum
import random


class Gender(Enum):
    MALE = 1
    FEMALE = 2


class BaseNameGenerator:
    @staticmethod
    def get_random_gender() -> Gender:
        if random.randint(1, 2) == 1:
            return Gender.MALE
        return Gender.FEMALE

    def generate_first_name(self, gender: Gender) -> str:
        pass

    def generate_surname(self, gender: Gender) -> str:
        pass

    def get_name(self)-> str:
        gender = self.get_random_gender()
        return self.generate_first_name(gender) + " " + self.generate_surname(gender)

File: Behavioral/template-method/examples/name-generator/C#/NameGenerator/NameGenerator/Generators/BaseNameGenerator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NameGenerator.Generators
{
    public enum Gender
    {
        Male,
        Female
    }

    public abstract class BaseNameGenerator
    {
        protected  Random _random = new Random();

        private Gender GetRandomGender()
        {
            switch(_random.Next(0, 2))
            {
                case 0:
                    return Gender.Male;
                case 1:
                    return Gender.Female;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }

        protected abstract string GenerateFirstName(Gender gender);

        protected abstract string GenerateSurname(Gender gender);

        public string GetName()
        {
            var gender = GetRandomGender();
            return $"{GenerateFirstName(gender)}  {GenerateSurname(gender)}";
        }
    }
}
